src/impl_/pyclass/assertions.rs

```diff
@@ -0,0 +1,52 @@
+/// Helper function that can be used at compile time to emit a diagnostic if
+/// the type does not implement `Sync` when it should.
+///
+/// The mere act of invoking this function will cause the diagnostic to be
+/// emitted if `T` does not implement `Sync` when it should.
+///
+/// The additional `const IS_SYNC: bool` parameter is used to allow the custom
+/// diagnostic to be emitted; if `PyClassSync`
+pub const fn assert_pyclass_sync<T, const IS_SYNC: bool>()
+where
+    T: PyClassSync<IS_SYNC> + Sync,
+{
+}
+
+#[cfg_attr(
+    diagnostic_namespace,
+    diagnostic::on_unimplemented(
+        message = "the trait `Sync` is not implemented for `{Self}`",
+        label = "needs to implement `Sync` to be `#[pyclass]`",
+        note = "to opt-out of threading support, use `#[pyclass(unsendable)]`",
```

Comment:
I'd prefer "replace non-Sync things by Sync things" - not sure on the wording for that.